### PR TITLE
cp2k: update to 4.1 + update of dependencies

### PR DESCRIPTION
**Pull request by alalazo**
##### Modifications
- [x] updated libint
- [x] updated wannier90
- [x] updated elpa
- [x] updated pexsi
- [x] updated cp2k
- [x] updated support for plumed
- [x] ported netlib-scalapack to CMakePackage

### CODE REVIEW

File: var/spack/repos/builtin/packages/libint/package.py [hunk 1]
```diff
@@ -25,25 +25,27 @@
 from spack import *
 
 
-class Libint(Package):
+class Libint(AutotoolsPackage):
     """Libint is a high-performance library for computing
-    Gaussian integrals in quantum mechanics."""
+    Gaussian integrals in quantum mechanics.
+    """
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/libint/package.py]: ??? Why this change?  It's not really consistent with the change to elpa

File: var/spack/repos/builtin/packages/libint/package.py [hunk 2]
```diff
@@ -25,25 +25,27 @@
...
     version('2.1.0', 'd0dcb985fe32ddebc78fe571ce37e2d6')
     version('1.1.6', '990f67b55f49ecc18f32c58da9240684')
     version('1.1.5', '379b7d0718ff398715d6898807adf628')
 
     # Build dependencies
     depends_on('autoconf@2.52:', type='build')
-    depends_on('automake',       type='build')
-    depends_on('libtool',        type='build')
+    depends_on('automake', type='build')
+    depends_on('libtool', type='build')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/libint/package.py]: PEP8 is silent on these spacing issues, and (AFAIK) Spack has not taken an official position on them.  As long as either way is OK, then changing it yields no objective improvement in the code (objective == Spack standards).  I believe we should avoid such edits in our PRs.  Not the least because they could result in "edit wars" in which certain lines keep getting changed back and forth, depending on the last person who edited them.  (For example, there are other PRs that end up ADDING spaces where this PR is REMOVING spaces).

File: var/spack/repos/builtin/packages/libint/package.py [hunk 3]
```diff
@@ -54,16 +56,14 @@ def url_for_version(self, version):
...
         aclocal('-I', 'lib/autoconf')
         autoconf()
 
-        config_args = [
-            '--prefix={0}'.format(prefix),
-            '--enable-shared'
-        ]
+    def configure_args(self):
+
+        cargs = ['--enable-shared']
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/libint/package.py]: Calling this `config_args` would both be clearer, and result in fewer changes to the code.

File: var/spack/repos/builtin/packages/libint/package.py (hunk 1 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/libint/package.py]: Please read [here](https://www.python.org/dev/peps/pep-0257)

File: var/spack/repos/builtin/packages/libint/package.py (hunk 2 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/libint/package.py]: It seems I have E241 and E242 turned on on `autopep8`. I can't say that I don't agree with them, but if this bothers people I can revert the changes.

**Comment by citibeth** [comment]
I agree with them.  As long as our flake8 program forces things in one
direction, I am happy with these changes.  I just hant to avoid edit wars.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 4]
```diff
@@ -48,13 +49,18 @@ class Cp2k(Package):
     depends_on('blas')
     depends_on('fftw')
     depends_on('libint@:1.2', when='@3.0')
+    depends_on('libint@:1.2', when='@4.1')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cp2k/package.py]: Why not just:
```python
depends_on('libint@:1.2')
```

File: var/spack/repos/builtin/packages/libint/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/libint/package.py]: I think @alalazo and I are the biggest perpetrators of the "edit wars". @alalazo uses `autopep8` to format everything consistently. This makes it easy to reformat a big package and ensure that it complies with `flake8`. I like to line things up because I think it makes it easier for humans to read. Neither of us are really "right". Enforcing `flake8` has made the "edit wars" much more benign than they used to be. If we could make flake8 even stricter it would prevent this, but I'm not going to vote for that because then I wouldn't be able to line things up anymore 😝 

Basically, I don't mind reviewing simple changes like this in a PR as long as others don't mind if I line things up once in a while.

File: var/spack/repos/builtin/packages/pexsi/package.py [hunk 5]
```diff
@@ -85,20 +95,25 @@ def install(self, spec, prefix):
...
         )
+
         # fortran "interface"
-        make('-C', 'fortran')
-        install_tree(
-            join_path(self.stage.source_path, 'fortran'),
-            join_path(self.prefix, 'fortran')
-        )
+        if '+fortran' in self.spec:
+            make('-C', 'fortran')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pexsi/package.py]: Should this line be in the `build` phase?

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 4 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/cp2k/package.py]: I can do it, but it seems `cp2k` web-site does not provide instructions to build old versions (or at least I can't find them). I opted then to have a per-version rule so that we can keep track of restrictions on the version of some dependencies. Maybe I can turn this into:
```python
depends_on('libint@:1.2', when='@3.0,4.1')
```
What do you think?

File: var/spack/repos/builtin/packages/libint/package.py (hunk 2 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/libint/package.py]: Come on, it's a friendly match not a war... :smile:

File: var/spack/repos/builtin/packages/pexsi/package.py (hunk 5 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/pexsi/package.py]: Well spotted

File: var/spack/repos/builtin/packages/pexsi/package.py [hunk 6]
```diff
@@ -85,20 +95,29 @@ def install(self, spec, prefix):
         for key, value in substitutions.items():
             filter_file(key, value, makefile)
 
-        make()
+    def build(self, spec, prefix):
+        super(Pexsi, self).build(spec, prefix)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pexsi/package.py]: I would just run `make()` here. The super class doesn't do anything fancy. I think it would be more obvious how things work if you did that.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 7]
```diff
@@ -47,14 +48,18 @@ class Cp2k(Package):
...
     depends_on('scalapack', when='+mpi')
     depends_on('plumed+shared+mpi', when='+plumed+mpi')
     depends_on('plumed+shared~mpi', when='+plumed~mpi')
-    depends_on('pexsi', when='+mpi')
-    depends_on('wannier90', when='+mpi')
+    depends_on('pexsi+fortran', when='+mpi')
+
+    # Apparently cp2k@4.1 needs an "experimental" version of libwannier.a
+    # which is only available contacting the developer directly. See INSTALL
+    # in the stage of cp2k@4.1
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cp2k/package.py]: Tell me a bit more about this. Does this mean that cp2k@4.1 can't be built with Spack? If so we should make the older version the default.

Comment (alalazo) [inline var/spack/repos/builtin/packages/cp2k/package.py]: No, it builds. The comment is just to explain why the dependency on `wannier90` is only for `cp2k@3.0`, while in principle also `cp2k@4.1` can be built with the optional `wannier90` support. 

The long story is that I got a symbol missing at linking time (`wannier_setup_simple`) and I had to inspect `cp2k` and `wannier90` sources to understand that `cp2k@4.1` was calling a function that was not in the `libwannier.a` library. Then I started looking into every file in `cp2k` and found the message:

> ==== 2p. libwannier90 (optional, interface to Wannier90 code) ====
>  * Needs an experimental version of wannier90 (currently only available from
>    hutter@chem.uzh.ch)
>  * Add -D__WANNIER90 to DFLAGS, -L$(LIBWANNIER90_DIR)-lwannier90 to LIBS.

hence the comment above.

File: var/spack/repos/builtin/packages/pexsi/package.py (hunk 6 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/pexsi/package.py]: Leaving it like this we can still take advantage of `build_targets` if need be.

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 7 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cp2k/package.py]: Ah, I see. Once the next version of `wannier90` is released we can go back and re-add support then. I'm surprised `cp2k` would release a version that doesn't support the latest version of `wannier90`.
